src/Tools/BuildValidator/BuildConstructor.cs

```diff
@@ -116,47 +180,70 @@ private static (CSharpCompilationOptions, CSharpParseOptions, Encoding) CreateCS
 
             var parseOptions = CSharpParseOptions.Default.WithLanguageVersion(langVersion)
                 .WithPreprocessorSymbols(preprocessorSymbols);
 
-            var (optimizationLevel, _) = GetOptimizationLevel(optimization);
+            var (optimizationLevel, plus) = GetOptimizationLevel(optimization);
 
             var nullableOptions = nullable is null
                 ? NullableContextOptions.Disable
                 : (NullableContextOptions)Enum.Parse(typeof(NullableContextOptions), nullable);
 
             var compilationOptions = new CSharpCompilationOptions(
-                pdbReader.GetOutputKind(),
+                optionsReader.GetOutputKind(),
                 reportSuppressedDiagnostics: false,
-                moduleName: null,
-                mainTypeName: null,
+
+                // PROTOTYPE: can't rely on the implicity moduleName here. In the case of .NET Core EXE the output name will
+                // end with .dll but the inferred name will be .exe
+                moduleName: assemblyName + ".dll",
+                mainTypeName: optionsReader.GetMainTypeName(),
                 scriptClassName: null,
                 usings: null,
                 optimizationLevel,
                 !string.IsNullOrEmpty(checkedString) && bool.Parse(checkedString),
                 !string.IsNullOrEmpty(unsafeString) && bool.Parse(unsafeString),
                 cryptoKeyContainer: null,
                 cryptoKeyFile: null,
-                cryptoPublicKey: default,
+                cryptoPublicKey: optionsReader.GetPublicKey()?.ToImmutableArray() ?? default,
                 delaySign: null,
                 Platform.AnyCpu,
+
+                // presence of diagnostics is expected to not affect emit.
                 ReportDiagnostic.Suppress,
                 warningLevel: 4,
                 specificDiagnosticOptions: null,
+
                 concurrentBuild: true,
                 deterministic: true,
+
                 xmlReferenceResolver: null,
                 sourceReferenceResolver: null,
                 metadataReferenceResolver: null,
+
                 assemblyIdentityComparer: null,
                 strongNameProvider: null,
                 publicSign: false,
+
                 metadataImportOptions: MetadataImportOptions.Public,
                 nullableContextOptions: nullableOptions);
+            compilationOptions.DebugPlusMode = plus;
 
             return (compilationOptions, parseOptions, encoding);
+
+            bool TryGetUniqueOption(string name, [NotNullWhen(true)] out string? value)
+            {
+                var result = pdbCompilationOptions.TryGetUniqueOption(name, out value);
+                _logger.LogInformation($"{name} - {value}");
+                return result;
+            }
+
+            string GetUniqueOption(string name)
+            {
+                _ = TryGetUniqueOption(name, out var value);
+                return value!;
```

Comment:
If the call to `TryGetUniqueOption` is expected to always succeed, consider adding an assertion or throwing an exception if it returned false.

Reply:
Fixed